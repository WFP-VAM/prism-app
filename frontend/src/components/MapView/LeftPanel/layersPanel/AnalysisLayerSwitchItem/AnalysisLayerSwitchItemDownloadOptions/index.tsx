import React, { memo, useMemo, useCallback, useState } from 'react';
import { IconButton, Menu, MenuItem, Tooltip } from '@material-ui/core';
import GetAppIcon from '@material-ui/icons/GetApp';
import { useSafeTranslation } from '../../../../../../i18n';
import { downloadToFile } from '../../../../utils';
import {
  BaselineLayerResult,
  downloadCSVFromTableData,
  ExposedPopulationResult,
  generateAnalysisFilename,
  PolygonAnalysisResult,
  useAnalysisTableColumns,
} from '../../../../../../utils/analysis-utils';

const AnalysisLayerSwitchItemDownloadOptions = memo(
  ({
    selected,
    analysisData,
    analysisResultSortByKey,
    analysisResultSortOrder,
  }: AnalysisLayerSwitchItemDownloadOptionsProps) => {
    const [
      downloadMenuAnchorEl,
      setDownloadMenuAnchorEl,
    ] = useState<HTMLElement | null>(null);

    const { translatedColumns } = useAnalysisTableColumns(analysisData);

    const { t } = useSafeTranslation();

    const featureCollection = useMemo(() => {
      return analysisData?.featureCollection;
    }, [analysisData]);

    const doesLayerAcceptCSVDownload = useMemo(() => {
      return (
        analysisData &&
        (analysisData instanceof BaselineLayerResult ||
          analysisData instanceof PolygonAnalysisResult)
      );
    }, [analysisData]);

    const handleDownloadMenuClose = useCallback(() => {
      setDownloadMenuAnchorEl(null);
    }, []);

    const handleDownloadMenuOpen = useCallback(
      (event: React.MouseEvent<HTMLElement>) => {
        setDownloadMenuAnchorEl(event.currentTarget);
      },
      [],
    );

    const renderedDownloadButton = useMemo(() => {
      if (!selected) {
        return (
          <IconButton disabled={!selected} onClick={handleDownloadMenuOpen}>
            <GetAppIcon />
          </IconButton>
        );
      }
      return (
        <Tooltip title="Download">
          <IconButton onClick={handleDownloadMenuOpen}>
            <GetAppIcon />
          </IconButton>
        </Tooltip>
      );
    }, [handleDownloadMenuOpen, selected]);

    const analysisDate = useMemo(() => {
      if (analysisData instanceof BaselineLayerResult) {
        return analysisData.analysisDate;
      }
      if (analysisData instanceof PolygonAnalysisResult) {
        return analysisData.endDate;
      }
      return null;
    }, [analysisData]);

    const fileName = useMemo(() => {
      if (
        // Explicit condition for type narrowing
        !analysisData ||
        analysisData instanceof ExposedPopulationResult
      ) {
        return analysisData?.getTitle(t);
      }
      return generateAnalysisFilename(analysisData, analysisDate ?? null);
    }, [analysisData, analysisDate, t]);

    const handleDownloadCsv = useCallback((): void => {
      if (
        // Explicit condition for type narrowing
        !analysisData ||
        analysisData instanceof ExposedPopulationResult
      ) {
        return;
      }
      downloadCSVFromTableData(
        analysisData,
        translatedColumns,
        analysisDate ?? null,
        analysisResultSortByKey,
        analysisResultSortOrder,
      );
    }, [
      analysisData,
      analysisDate,
      analysisResultSortByKey,
      analysisResultSortOrder,
      translatedColumns,
    ]);

    const handleDownloadGeoJson = useCallback(() => {
      if (!analysisData) {
        return;
      }
      downloadToFile(
        {
          content: JSON.stringify(featureCollection),
          isUrl: false,
        },
        fileName ?? 'prism_extract',
        'application/json',
      );
    }, [analysisData, featureCollection, fileName]);

    const renderedDownloadAsCSVMenuItem = useMemo(() => {
      if (!doesLayerAcceptCSVDownload) {
        return null;
      }
      return (
        <MenuItem key="download-as-csv" onClick={handleDownloadCsv}>
          {t('Download as CSV')}
        </MenuItem>
      );
    }, [doesLayerAcceptCSVDownload, handleDownloadCsv, t]);

    return (
      <>
        {renderedDownloadButton}
        <Menu
          id="download-menu"
          anchorEl={downloadMenuAnchorEl}
          keepMounted
          open={Boolean(downloadMenuAnchorEl)}
          onClose={handleDownloadMenuClose}
        >
          {[
            renderedDownloadAsCSVMenuItem,
            <MenuItem key="download-as-geojson" onClick={handleDownloadGeoJson}>
              {t('Download as GeoJSON')}
            </MenuItem>,
          ]}
        </Menu>
      </>
    );
  },
);

interface AnalysisLayerSwitchItemDownloadOptionsProps {
  selected: boolean;
  analysisData?:
    | BaselineLayerResult
    | PolygonAnalysisResult
    | ExposedPopulationResult;
  analysisResultSortByKey: string | number;
  analysisResultSortOrder: 'asc' | 'desc';
}

export default AnalysisLayerSwitchItemDownloadOptions;
