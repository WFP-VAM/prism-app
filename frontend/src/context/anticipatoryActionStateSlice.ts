import { PayloadAction, createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import Papa from 'papaparse';
import { LayerDefinitions } from 'config/utils';
import { AnticipatoryActionLayerProps } from 'config/types';
import type { CreateAsyncThunkTypes, RootState } from './store';

export const allWindowsKey = 'All';

// na/ny are not actually found in CSV, but defined not to cause confusion when calling the functions
export const AAcategory = ['na', 'ny', 'Leve', 'Moderado', 'Severo'] as const;
export type AACategoryType = typeof AAcategory[number];

export const AAPhase = ['na', 'ny', 'Ready', 'Set'] as const;
export type AAPhaseType = typeof AAPhase[number];

export interface AnticipatoryActionData {
  category: AACategoryType;
  district: string;
  index: string;
  month: string;
  phase: AAPhaseType;
  probability: string;
  trigger: string;
  triggerNB: string;
  triggerType: string;
  type: string;
  windows: string;
  yearOfIssue: string;
  date: string;
}

export const AAlayerKey = 'anticipatory_action';

const AACSVKeys: [string, string][] = [
  ['Category', 'category'],
  ['District', 'district'],
  ['Index', 'index'],
  ['Month', 'month'],
  ['Phase', 'phase'],
  ['Probability', 'probability'],
  ['Trigger', 'trigger'],
  ['Trigger_nb', 'triggerNB'],
  ['Trigger_type', 'triggerType'],
  ['Type', 'type'],
  ['Windows', 'windows'],
  ['Year_of_issue', 'yearOfIssue'],
];

function transform(data: any[], keys: [string, string][]) {
  const parsed = data
    .filter(x => !!x[keys[0][0]]) // filter empty rows
    .map(obj => {
      const entries = keys.map(k => [k[1], obj[k[0]]]);
      const month = obj.Month.padStart(2, '0');
      const year = obj.Year_of_issue.split('-')[0];

      // TODO: update this once CSV date format is revised
      const actualYear = Number(month) < 5 ? String(Number(year) + 1) : year;
      const date = `${actualYear}-${month}-01`;
      return Object.fromEntries([...entries, ['date', date]]);
    }) as AnticipatoryActionData[];

  const windows = [...new Set(parsed.map(x => x.windows))];
  // eslint-disable-next-line fp/no-mutating-methods
  const availableDates = [
    ...new Set(parsed.map(x => new Date(x.date).getTime())),
  ].sort();

  const groupedByDistrict = new Map<string, AnticipatoryActionData[]>();

  parsed.forEach(x => {
    const { district } = x;
    const rows = groupedByDistrict.get(district);
    groupedByDistrict.set(district, rows ? [...rows, x] : [x]);
  });

  const sortFn = (a: AnticipatoryActionData, b: AnticipatoryActionData) => {
    const aDate = new Date(a.date).valueOf();
    const bDate = new Date(b.date).valueOf();
    if (aDate > bDate) {
      return -1;
    }
    if (bDate > aDate) {
      return 1;
    }
    const aCatIndex = AAcategory.findIndex(x => x === a.category);
    const bCatIndex = AAcategory.findIndex(x => x === b.category);
    if (aCatIndex > bCatIndex) {
      return -1;
    }
    if (bCatIndex > aCatIndex) {
      return 1;
    }
    return 0;
  };

  const result = Object.fromEntries(
    // eslint-disable-next-line fp/no-mutating-methods
    Array.from(groupedByDistrict.entries()).map(x => [x[0], x[1].sort(sortFn)]),
  );

  const monitoredDistricts = Object.keys(result);

  return { data: result, windows, availableDates, monitoredDistricts };
}

type AnticipatoryActionState = {
  data: {
    [k: string]: AnticipatoryActionData[];
  };
  availableDates?: number[];
  monitoredDistricts: string[];
  windows: string[];
  selectedWindow: string;
  categoryFilters: Record<AACategoryType, boolean>;
  loading: boolean;
  error: string | null;
};

const initialState: AnticipatoryActionState = {
  data: {},
  windows: [],
  availableDates: undefined,
  monitoredDistricts: [],
  selectedWindow: allWindowsKey,
  categoryFilters: {
    Severo: true,
    Moderado: true,
    Leve: true,
    na: false,
    ny: false,
  },
  loading: false,
  error: null,
};

export const loadAAData = createAsyncThunk<
  {
    data: {
      [k: string]: AnticipatoryActionData[];
    };
    windows: string[];
    availableDates: number[];
    monitoredDistricts: string[];
  },
  undefined,
  CreateAsyncThunkTypes
>('anticipatoryActionState/loadAAData', async () => {
  const layer = LayerDefinitions[AAlayerKey] as AnticipatoryActionLayerProps;

  const url = layer.baseUrl;
  return new Promise<any>((resolve, reject) =>
    Papa.parse(url, {
      header: true,
      download: true,
      complete: results => resolve(transform(results.data, AACSVKeys)),
      error: error => reject(error),
    }),
  );
});

export const anticipatoryActionStateSlice = createSlice({
  name: 'anticipatoryActionState',
  initialState,
  reducers: {
    setSelectedWindow: (state, { payload }: PayloadAction<string>) => ({
      ...state,
      selectedWindow: payload,
    }),
    setCategoryFilters: (
      state,
      {
        payload,
      }: PayloadAction<
        Partial<Record<'Leve' | 'Moderado' | 'Severo', boolean>>
      >,
    ) => ({
      ...state,
      categoryFilters: { ...state.categoryFilters, ...payload },
    }),
  },
  extraReducers: builder => {
    builder.addCase(loadAAData.fulfilled, (state, { payload }) => ({
      ...state,
      loading: false,
      data: payload.data,
      windows: payload.windows,
      availableDates: payload.availableDates,
      monitoredDistricts: payload.monitoredDistricts,
    }));

    builder.addCase(loadAAData.rejected, (state, action) => ({
      ...state,
      loading: false,
      error: action.error.message
        ? action.error.message
        : action.error.toString(),
    }));

    builder.addCase(loadAAData.pending, ({ error, ...state }) => ({
      ...state,
      error: null,
      loading: true,
    }));
  },
});

// export selectors
export const AnticipatoryActionDataSelector = (state: RootState) =>
  state.anticipatoryActionState.data;

export const AnticipatoryActionWindowsSelector = (state: RootState) =>
  state.anticipatoryActionState.windows;

export const AnticipatoryActionAvailableDatesSelector = (state: RootState) =>
  state.anticipatoryActionState.availableDates;

export const AASelectedWindowSelector = (state: RootState) =>
  state.anticipatoryActionState.selectedWindow;

export const AACategoryFiltersSelector = (state: RootState) =>
  state.anticipatoryActionState.categoryFilters;

// export actions
export const {
  setSelectedWindow,
  setCategoryFilters,
} = anticipatoryActionStateSlice.actions;

export default anticipatoryActionStateSlice.reducer;
