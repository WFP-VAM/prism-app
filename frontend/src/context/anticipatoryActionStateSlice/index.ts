import { PayloadAction, createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import Papa from 'papaparse';
import { DateItem } from 'config/types';
import { appConfig } from 'config';
import { AAWindowKeys } from 'config/utils';
import type { CreateAsyncThunkTypes, RootState } from '../store';
import {
  AACategoryType,
  AAView,
  AnticipatoryActionData,
  AnticipatoryActionState,
  Vulnerability,
  allWindowsKey,
} from './types';
import {
  calculateMapRenderedDistricts,
  emptyWindows,
  parseAndTransformAA,
} from './utils';

const initialState: AnticipatoryActionState = {
  data: emptyWindows,
  availableDates: undefined,
  monitoredDistricts: [],
  filters: {
    selectedDate: undefined,
    selectedWindow: allWindowsKey,
    selectedIndex: '',
    categories: {
      Severe: true,
      Moderate: true,
      Mild: true,
      Normal: true,
      na: true,
      ny: true,
    },
  },
  markers: [],
  selectedDistrict: '',
  renderedDistricts: emptyWindows,
  windowRanges: { 'Window 1': undefined, 'Window 2': undefined },
  view: AAView.Home,
  loading: false,
  error: null,
};

export const loadAAData = createAsyncThunk<
  {
    windowData: {
      data: AnticipatoryActionData;
      availableDates: DateItem[];
      windowKey: (typeof AAWindowKeys)[number];
      range: { start: string; end: string };
    }[];
    monitoredDistricts: { name: string; vulnerability: Vulnerability }[];
  },
  undefined,
  CreateAsyncThunkTypes
>('anticipatoryActionState/loadAAData', async () => {
  const url = appConfig.anticipatoryActionUrl;

  return new Promise<any>((resolve, reject) => {
    Papa.parse(url, {
      header: true,
      download: true,
      complete: results => resolve(parseAndTransformAA(results.data)),
      error: error => reject(error),
    });
  });
});

export const anticipatoryActionStateSlice = createSlice({
  name: 'anticipatoryActionState',
  initialState,
  reducers: {
    setAAFilters: (
      state,
      {
        payload,
      }: PayloadAction<
        Partial<{
          selectedDate: string | undefined;
          selectedWindow: (typeof AAWindowKeys)[number] | typeof allWindowsKey;
          selectedIndex: string;
          categories: Partial<Record<AACategoryType, boolean>>;
        }>
      >,
    ) => {
      const { categories, ...rest } = payload;
      const newCategories =
        categories !== undefined
          ? { ...state.filters.categories, ...categories }
          : state.filters.categories;
      const newFilters = {
        ...state.filters,
        ...rest,
        categories: newCategories,
      };
      return {
        ...state,
        filters: newFilters,
        renderedDistricts: calculateMapRenderedDistricts({
          filters: newFilters,
          data: state.data,
          windowRanges: state.windowRanges,
        }),
      };
    },
    setAASelectedDistrict: (
      state,
      { payload }: PayloadAction<AnticipatoryActionState['selectedDistrict']>,
    ) => ({
      ...state,
      selectedDistrict: payload,
    }),
    setAAMarkers: (
      state,
      { payload }: PayloadAction<AnticipatoryActionState['markers']>,
    ) => ({
      ...state,
      markers: payload,
    }),
    setAAView: (
      state,
      { payload }: PayloadAction<AnticipatoryActionState['view']>,
    ) => ({
      ...state,
      view: payload,
    }),
  },
  extraReducers: builder => {
    builder.addCase(loadAAData.fulfilled, (state, { payload }) => {
      const newData = Object.fromEntries(
        payload.windowData.map(x => [x.windowKey, x.data]),
      ) as Record<(typeof AAWindowKeys)[number], AnticipatoryActionData>;
      const newRanges = Object.fromEntries(
        payload.windowData.map(x => [x.windowKey, x.range]),
      ) as Record<
        (typeof AAWindowKeys)[number],
        { start: string; end: string }
      >;
      return {
        ...state,
        loading: false,
        data: newData,
        availableDates: Object.fromEntries(
          payload.windowData.map(x => [x.windowKey, x.availableDates]),
        ) as Record<(typeof AAWindowKeys)[number], DateItem[]>,
        windowRanges: newRanges,
        monitoredDistricts: payload.monitoredDistricts,
        renderedDistricts: calculateMapRenderedDistricts({
          filters: state.filters,
          data: newData,
          windowRanges: newRanges,
        }),
      };
    });

    builder.addCase(loadAAData.rejected, (state, action) => ({
      ...state,
      loading: false,
      error: action.error.message
        ? action.error.message
        : action.error.toString(),
    }));

    builder.addCase(loadAAData.pending, ({ error: _error, ...state }) => ({
      ...state,
      error: null,
      loading: true,
    }));
  },
});

// export selectors
export const AADataSelector = (state: RootState) =>
  state.anticipatoryActionState.data;

export const AAAvailableDatesSelector = (state: RootState) =>
  state.anticipatoryActionState.availableDates;

export const AAMonitoredDistrictsSelector = (state: RootState) =>
  state.anticipatoryActionState.monitoredDistricts;

export const AAFiltersSelector = (state: RootState) =>
  state.anticipatoryActionState.filters;

export const AARenderedDistrictsSelector = (state: RootState) =>
  state.anticipatoryActionState.renderedDistricts;

export const AASelectedDistrictSelector = (state: RootState) =>
  state.anticipatoryActionState.selectedDistrict;

export const AAMarkersSelector = (state: RootState) =>
  state.anticipatoryActionState.markers;

export const AAViewSelector = (state: RootState) =>
  state.anticipatoryActionState.view;

export const AAWindowRangesSelector = (state: RootState) =>
  state.anticipatoryActionState.windowRanges;

// export actions
export const { setAAFilters, setAASelectedDistrict, setAAMarkers, setAAView } =
  anticipatoryActionStateSlice.actions;

export default anticipatoryActionStateSlice.reducer;
