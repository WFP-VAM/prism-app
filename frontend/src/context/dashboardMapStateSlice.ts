import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { Map as MaplibreMap, LngLatBounds } from 'maplibre-gl';

// Maplibre's map type contains some kind of cyclic dependency that causes an infinite loop in immers's change
// tracking. To save it off, we wrap it in a JS closure so that Redux just checks the function for changes, rather
// than recursively walking the whole object.
type MapGetter = () => MaplibreMap | undefined;

export interface DashboardMapState {
  maplibreMap: MapGetter;
  layers: any[];
  bounds: LngLatBounds | null;
  zoom: number | null;
}

const initialState: DashboardMapState = {
  maplibreMap: (() => {}) as MapGetter,
  layers: [],
  bounds: null,
  zoom: null,
};

export const dashboardMapStateSlice = createSlice({
  name: 'dashboardMapState',
  initialState,
  reducers: {
    setDashboardMap: (state, action: PayloadAction<MapGetter>) => ({
      ...state,
      maplibreMap: action.payload,
    }),
    addDashboardLayer: (state, action: PayloadAction<any>) => ({
      ...state,
      layers: [...state.layers, action.payload],
    }),
    removeDashboardLayer: (state, action: PayloadAction<string>) => ({
      ...state,
      layers: state.layers.filter(layer => layer.id !== action.payload),
    }),
    setDashboardLayers: (state, action: PayloadAction<any[]>) => ({
      ...state,
      layers: action.payload,
    }),
    setDashboardBounds: (state, action: PayloadAction<LngLatBounds>) => ({
      ...state,
      bounds: action.payload,
    }),
    setDashboardLocation: (
      state,
      action: PayloadAction<{ bounds: LngLatBounds; zoom: number }>,
    ) => ({
      ...state,
      bounds: action.payload.bounds,
      zoom: action.payload.zoom,
    }),
  },
});

// Getters
export const dashboardMapSelector = (state: {
  dashboardMapState: DashboardMapState;
}): MaplibreMap | undefined => state.dashboardMapState.maplibreMap();

export const dashboardLayersSelector = (state: {
  dashboardMapState: DashboardMapState;
}): any[] => state.dashboardMapState.layers;

export const dashboardBoundsSelector = (state: {
  dashboardMapState: DashboardMapState;
}): LngLatBounds | null => state.dashboardMapState.bounds;

export const dashboardZoomSelector = (state: {
  dashboardMapState: DashboardMapState;
}): number | null => state.dashboardMapState.zoom;

// Setters
export const {
  setDashboardMap,
  addDashboardLayer,
  removeDashboardLayer,
  setDashboardLayers,
  setDashboardBounds,
  setDashboardLocation,
} = dashboardMapStateSlice.actions;

export default dashboardMapStateSlice.reducer;
