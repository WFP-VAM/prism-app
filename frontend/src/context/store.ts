import {
  combineReducers,
  configureStore,
  getDefaultMiddleware,
} from '@reduxjs/toolkit';
import mapStateReduce from './mapStateSlice';
import serverStateReduce from './serverStateSlice';
import tableStateReduce from './tableStateSlice';
import tooltipStateReduce from './tooltipStateSlice';
import notificationStateReduce, {
  errorToNotificationMiddleware,
} from './notificationStateSlice';
import analysisResultStateReduce from './analysisResultStateSlice';
import mapSelectionLayerStateReduce from './mapSelectionLayerStateSlice';
import mapBoundaryInfoStateReduce from './mapBoundaryInfoStateSlice';
import datasetResultStateReduce from './datasetStateSlice';
import mapTileLoadingStateReduce from './mapTileLoadingStateSlice';
import leftPanelStateReduce from './leftPanelStateSlice';
import opacityStateReduce from './opacityStateSlice';
import anticipatoryActionDroughtStateReduce from './anticipatoryAction/AADroughtStateSlice';
import anticipatoryActionStormStateReduce from './anticipatoryAction/AAStormStateSlice';

const reducer = combineReducers({
  mapState: mapStateReduce,
  serverState: serverStateReduce,
  tableState: tableStateReduce,
  tooltipState: tooltipStateReduce,
  analysisResultState: analysisResultStateReduce,
  notificationState: notificationStateReduce,
  mapSelectionLayerStateSlice: mapSelectionLayerStateReduce,
  mapBoundaryInfoState: mapBoundaryInfoStateReduce,
  datasetState: datasetResultStateReduce,
  mapTileLoadingState: mapTileLoadingStateReduce,
  leftPanelState: leftPanelStateReduce,
  opacityState: opacityStateReduce,
  anticipatoryActionDroughtState: anticipatoryActionDroughtStateReduce,
  anticipatoryActionStormState: anticipatoryActionStormStateReduce,
});

export const store = configureStore({
  reducer,
  middleware: getDefaultMiddleware({
    // TODO: Instead of snoozing this check, we might want to
    // serialize the state
    serializableCheck: false,
    immutableCheck: {
      // do not check the following state branches for accidental
      // mutations. This saves a lot of time in dev mode (but has no
      // impact on production builds).
      ignoredPaths: [
        'mapState.boundaryRelations',
        'mapState.layersData',
        'analysisResultState.result',
        'serverState.availableDates',
      ],
    },
  }).concat(errorToNotificationMiddleware),
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof reducer>;

export type ThunkApi = {
  dispatch: AppDispatch;
  getState: () => RootState;
  // can add 'extra' and 'rejectValue' here if we ever need them
  // https://redux-toolkit.js.org/usage/usage-with-typescript#createasyncthunk
};

// Used as the third type definition for typing createAsyncThunk
// e.g. createAsyncThunk<ReturnType, HandlerType, CreateAsyncThunkTypes>
export type CreateAsyncThunkTypes = {
  dispatch: AppDispatch;
  state: RootState;
};
