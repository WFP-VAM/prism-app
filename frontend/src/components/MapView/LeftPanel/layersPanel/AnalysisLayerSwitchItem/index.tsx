import React, {
  memo,
  useState,
  useCallback,
  useMemo,
  ChangeEvent,
  useEffect,
} from 'react';
import {
  Box,
  IconButton,
  Slider,
  Switch,
  Tooltip,
  Typography,
} from '@material-ui/core';
import { createStyles, WithStyles, withStyles } from '@material-ui/styles';
import { useDispatch, useSelector } from 'react-redux';
import OpacityIcon from '@material-ui/icons/Opacity';
import {
  clearAnalysisResult,
  setAnalysisLayerOpacity,
} from '../../../../../context/analysisResultStateSlice';
import { handleChangeOpacity } from '../../../Legends/handleChangeOpacity';
import { mapSelector } from '../../../../../context/mapStateSlice/selectors';
import AnalysisLayerSwitchItemDownloadOptions from './AnalysisLayerSwitchItemDownloadOptions';
import {
  BaselineLayerResult,
  ExposedPopulationResult,
  PolygonAnalysisResult,
} from '../../../../../utils/analysis-utils';

const AnalysisLayerSwitchItem = memo(
  ({
    classes,
    title,
    initialOpacity,
    analysisData,
    analysisResultSortOrder,
    analysisResultSortByKey,
  }: AnalysisLayerSwitchItemProps) => {
    const dispatch = useDispatch();
    const map = useSelector(mapSelector);
    const [selected, setSelected] = useState<boolean>(true);
    const [isOpacitySelected, setIsOpacitySelected] = useState<boolean>(false);
    const [opacity, setOpacityValue] = useState<number>(initialOpacity || 0);

    useEffect(() => {
      setOpacityValue(initialOpacity || 0);
    }, [initialOpacity]);

    const handleOnChangeSwitch = useCallback(() => {
      setSelected(!selected);
      // The toggle button will be by default toggled
      dispatch(clearAnalysisResult());
    }, [dispatch, selected]);

    const handleOpacityClick = useCallback(() => {
      setIsOpacitySelected(!isOpacitySelected);
    }, [isOpacitySelected]);

    const handleChangeOpacityValue = useCallback(
      val => {
        setOpacityValue(val);
        dispatch(setAnalysisLayerOpacity(val));
      },
      [dispatch],
    );

    const handleOnChangeSliderValue = useCallback(
      (event: ChangeEvent<{}>, newValue: number | number[]) => {
        handleChangeOpacity(
          event,
          newValue as number,
          map,
          undefined, // We have to pass undefined here so that the function know that we are in an analysis custom layer
          undefined, // We have also to pass undefined here to show that we have analysis custom layer,
          handleChangeOpacityValue,
        );
      },
      [handleChangeOpacityValue, map],
    );

    const renderedOpacitySlider = useMemo(() => {
      if (!selected || !isOpacitySelected) {
        return null;
      }
      return (
        <Box display="flex" justifyContent="right" alignItems="center">
          <Box pr={3}>
            <Typography
              classes={{ root: classes.opacityText }}
            >{`Opacity ${Math.round(opacity * 100)}%`}</Typography>
          </Box>
          <Box width="25%" pr={3}>
            <Slider
              value={opacity}
              step={0.01}
              min={0}
              max={1}
              aria-labelledby="left-opacity-slider"
              classes={{
                root: classes.opacitySliderRoot,
                thumb: classes.opacitySliderThumb,
              }}
              onChange={handleOnChangeSliderValue}
            />
          </Box>
        </Box>
      );
    }, [
      classes.opacitySliderRoot,
      classes.opacitySliderThumb,
      classes.opacityText,
      handleOnChangeSliderValue,
      isOpacitySelected,
      opacity,
      selected,
    ]);

    const renderedOpacityIconButton = useMemo(() => {
      if (!selected) {
        return (
          <IconButton
            disabled={!selected}
            classes={{
              root: isOpacitySelected
                ? classes.opacityRootSelected
                : classes.opacityRoot,
            }}
            onClick={handleOpacityClick}
          >
            <OpacityIcon />
          </IconButton>
        );
      }
      return (
        <Tooltip title="Opacity">
          <IconButton
            disabled={!selected}
            classes={{
              root: isOpacitySelected
                ? classes.opacityRootSelected
                : classes.opacityRoot,
            }}
            onClick={handleOpacityClick}
          >
            <OpacityIcon />
          </IconButton>
        </Tooltip>
      );
    }, [
      classes.opacityRoot,
      classes.opacityRootSelected,
      handleOpacityClick,
      isOpacitySelected,
      selected,
    ]);

    return (
      <Box
        className={classes.analysisLayerSwitchRootItem}
        display="flex"
        flexDirection="column"
        maxWidth="100%"
      >
        <Box display="flex" alignItems="center" m={2}>
          <Switch
            size="small"
            className={classes.switch}
            classes={{
              switchBase: classes.switchBase,
              track: classes.switchTrack,
            }}
            checked={selected}
            onChange={handleOnChangeSwitch}
            inputProps={{
              'aria-label': title,
            }}
          />
          <Typography
            className={selected ? classes.title : classes.titleUnchecked}
          >
            {title}
          </Typography>
          {renderedOpacityIconButton}
          <AnalysisLayerSwitchItemDownloadOptions
            analysisData={analysisData}
            analysisResultSortByKey={analysisResultSortByKey}
            analysisResultSortOrder={analysisResultSortOrder}
            selected={selected}
          />
        </Box>
        {renderedOpacitySlider}
      </Box>
    );
  },
);

const styles = () =>
  createStyles({
    analysisLayerSwitchRootItem: {
      backgroundColor: '#FFFFFF',
    },
    title: {
      lineHeight: 1.8,
      color: 'black',
      fontWeight: 400,
    },
    titleUnchecked: {
      lineHeight: 1.8,
      color: '#828282',
      fontWeight: 400,
    },
    select: {
      '&::before': {
        border: 'none',
      },
    },
    selectItem: {
      whiteSpace: 'normal',
      fontSize: 13,
      fontWeight: 300,
      color: 'black',
      padding: 0,
      marginLeft: 5,
    },
    selectItemUnchecked: {
      whiteSpace: 'normal',
      fontSize: 13,
      fontWeight: 300,
      color: '#828282',
      padding: 0,
      marginLeft: 5,
    },
    switch: {
      marginRight: 2,
    },
    switchTrack: {
      backgroundColor: '#E0E0E0',
    },
    switchBase: {
      color: '#E0E0E0',
      '&.Mui-checked': {
        color: '#53888F',
      },
      '&.Mui-checked + .MuiSwitch-track': {
        backgroundColor: '#B1D6DB',
      },
    },
    opacityRoot: {
      color: '#828282',
      marginLeft: 'auto',
    },
    opacityRootSelected: {
      backgroundColor: '#4CA1AD',
      color: '#F2F2F2',
      marginLeft: 'auto',
      '&:hover': {
        color: '#4CA1AD',
      },
    },
    opacityText: {
      color: '#4CA1AD',
      marginBottom: '10px',
    },
    opacitySliderRoot: {
      color: '#4CA1AD',
      height: 8,
    },
    opacitySliderThumb: {
      backgroundColor: '#4CA1AD',
    },
  });

interface AnalysisLayerSwitchItemProps extends WithStyles<typeof styles> {
  title: string;
  initialOpacity: number;
  analysisData?:
    | BaselineLayerResult
    | PolygonAnalysisResult
    | ExposedPopulationResult;
  analysisResultSortByKey: string | number;
  analysisResultSortOrder: 'asc' | 'desc';
}

export default withStyles(styles)(AnalysisLayerSwitchItem);
