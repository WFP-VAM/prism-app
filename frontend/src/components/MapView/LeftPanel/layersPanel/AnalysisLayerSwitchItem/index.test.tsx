import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import React from 'react';
import { store } from '../../../../../context/store';
import AnalysisLayerSwitchItem from '.';

test('renders as expected', () => {
  const { container } = render(
    <Provider store={store}>
      <AnalysisLayerSwitchItem
        initialOpacity={0.5}
        analysisResultSortByKey="some sort by key"
        analysisResultSortOrder="asc"
        title="Some Analysis Layer Title"
      />
    </Provider>,
  );
  expect(container).toMatchSnapshot();
});
