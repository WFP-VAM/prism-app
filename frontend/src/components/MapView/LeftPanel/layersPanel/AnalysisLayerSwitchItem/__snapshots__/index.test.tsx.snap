// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-15 memo-analysisLayerSwitchRootItem-1"
  >
    <div
      class="MuiBox-root MuiBox-root-16"
    >
      <mock-switch
        checked="true"
        classes="[object Object]"
        classname="memo-switch-7"
        inputprops="[object Object]"
        size="small"
      />
      <mock-typography
        classname="memo-title-2"
      >
        Some Analysis Layer Title
      </mock-typography>
      <mock-tooltip
        title="Opacity"
      >
        <span>
          <button
            class="MuiButtonBase-root MuiIconButton-root memo-opacityRoot-10"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M17.66 8L12 2.35 6.34 8C4.78 9.56 4 11.64 4 13.64s.78 4.11 2.34 5.67 3.61 2.35 5.66 2.35 4.1-.79 5.66-2.35S20 15.64 20 13.64 19.22 9.56 17.66 8zM6 14c.01-2 .62-3.27 1.76-4.4L12 5.27l4.24 4.38C17.38 10.77 17.99 12 18 14H6z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </span>
      </mock-tooltip>
      <mock-tooltip
        title="Download"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 9h-4V3H9v6H5l7 7 7-7zM5 18v2h14v-2H5z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </mock-tooltip>
      <mock-menu
        id="download-menu"
        keepmounted="true"
        open="false"
      >
        <mock-menuitem>
          Download as CSV
        </mock-menuitem>
        <mock-menuitem>
          Download as GeoJSON
        </mock-menuitem>
      </mock-menu>
    </div>
  </div>
</div>
`;
